### PR TITLE
[inductor] Handle nested tuple/list output in fallback kernel

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88495

Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

### CODE REVIEW

**Comment by desertfire** [comment]
This is to fix an internal case. I found it hard to create a unit test for this in OSS.

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
...
-            packed = FallbackKernel(
-                MultiOutputLayout(tensor_args[0].get_device()),
-                kernel,
-                tensor_args,
-                non_tensor_args,
-                unflatten_args,
-            )
-            return [
-                (
+        def generate_output(output, index=""):
```
Comment (eellison) [inline torch/_inductor/ir.py:2920]: could add asserts that the length of the output at runtime is the same as we observed here ?

**Review by eellison — COMMENTED** [review]
looking good, one comment. For test maybe `aten::unique_dim` ?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
...
+            if isinstance(output, list):
```
Comment (ngimel) [inline torch/_inductor/ir.py:2921]: nit
```
if isinstance(output, (list, tuple)):
    return type(output)(...)
```

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
...
+                    generate_output(output[i], f"{index}[{i}]")
+                    for i in range(len(output))
+                )
+            elif isinstance(output, tuple):
+                return tuple(
+                    generate_output(output[i], f"{index}[{i}]")
+                    for i in range(len(output))
+                )
+            else:
+                return (
```
Comment (ngimel) [inline torch/_inductor/ir.py:2932]: should you assert on `output` type here, before trying to access its device and dtype?

File: torch/_inductor/ir.py (hunk 1 above)
Comment (desertfire) [reply torch/_inductor/ir.py:2920]: We do have `assert_size_stride` generated and that will trigger an issue if the output indexing is OOB (that's how I debugged one test failure), so I guess I will skip adding another assert for now.

Comment (eellison) [reply torch/_inductor/ir.py:2920]: that'll catch when the list runtime list is smaller but it won't catch when the runtime list is larger

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**